### PR TITLE
[Enhancement] Support padding in testing

### PR DESCRIPTION
**Pull request by xiexinch**
## Motivation

- Fix removing gt_sem_seg padding at postprocess_result
- Support  different padding methods in testing

## Modification

- Add removing gt_sem_seg padding at postprocess_result method in mmseg/models/segmentors/base.py.
- Add test_cfg` params to SegDataPreprocessor, supporting different padding methods.

## BC-breaking (Optional)

None

## Use cases

In a case that uses `size` in training and using `size_divisor` in testing, we might set the `test_cfg` param to data_preprocessor.

```python
data_preprocessor = dict(
    type='SegDataPreProcessor',
    mean=[123.675, 116.28, 103.53],
    std=[58.395, 57.12, 57.375],
    bgr_to_rgb=True,
    pad_val=0,
    seg_pad_val=255,
    size=(512, 512),
    test_cfg=dict(size_divisor=32))
```

### CODE REVIEW

File: mmseg/models/data_preprocessor.py [hunk 1]
```diff
@@ -48,18 +48,28 @@ class SegDataPreProcessor(BaseDataPreprocessor):
         rgb_to_bgr (bool): whether to convert image from RGB to RGB.
             Defaults to False.
         batch_augments (list[dict], optional): Batch-level augmentations
+        train_cfg (dict, optional): The padding size config in training, if
+            not specify, will use `size` and `size_divisor` params as default.
+            Defaults to None, only supports keys `size` or `size_divisor`.
+        test_cfg (dict, optional): The padding size config in testing, if not
+            specify, will use `size` and `size_divisor` params as default.
+            Defaults to None, only supports keys `size` or `size_divisor`.
```
Comment (MeowZheng) [inline mmseg/models/data_preprocessor.py:56]: Please add an example in PR description for `train_cfg` and `test_cfg` as a padding config for input data.

**Review by MeowZheng — COMMENTED** [review]
add ut for the new  function

**Review by MeowZheng — COMMENTED** [review]
please try to use `Pad` in test pipeline

**Comment by xiexinch** [comment]
> please try to use `Pad` in test pipeline

If use `Pad` in test _pipeline, the padded pixels would be normalized.
https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/data_preprocessor.py#L108

**Review by MeowZheng — APPROVED** [review]

File: mmseg/models/data_preprocessor.py [hunk 2]
```diff
@@ -122,10 +131,20 @@ def forward(self, data: dict, training: bool = False) -> Dict[str, Any]:
...
-                inputs=torch.stack(inputs, dim=0), data_samples=data_samples)
+            # pad images when testing
+            if self.test_cfg:
+                inputs, data_samples = stack_batch(
+                    inputs=inputs,
+                    data_samples=data_samples,
+                    size=self.test_cfg.get('size', None),
+                    size_divisor=self.test_cfg.get('size_divisor', None),
+                    pad_val=self.pad_val,
+                    seg_pad_val=self.seg_pad_val)
```
Comment (MeowZheng) [inline mmseg/models/data_preprocessor.py:146]:
```suggestion
                inputs, data_samples = stack_batch(
                    inputs=inputs,
                    size=self.test_cfg.get('size', None),
                    size_divisor=self.test_cfg.get('size_divisor', None),
                    pad_val=self.pad_val,
                    seg_pad_val=self.seg_pad_val)
```

File: mmseg/models/segmentors/base.py [hunk 3]
```diff
@@ -165,6 +165,12 @@ def postprocess_result(self,
                 i_seg_logits = seg_logits[i:i + 1, :,
                                           padding_top:H - padding_bottom,
                                           padding_left:W - padding_right]
+                i_gt_sem_seg = data_samples[i].gt_sem_seg.data[:,
+                                                               padding_top:H -
+                                                               padding_bottom,
+                                                               padding_left:W -
+                                                               padding_right]
+
```
Comment (MeowZheng) [inline mmseg/models/segmentors/base.py:173]:
```suggestion
```

File: mmseg/models/segmentors/base.py [hunk 4]
```diff
@@ -184,7 +190,10 @@ def postprocess_result(self,
                 'seg_logits':
                 PixelData(**{'data': i_seg_logits}),
                 'pred_sem_seg':
-                PixelData(**{'data': i_seg_pred})
+                PixelData(**{'data': i_seg_pred}),
+                'gt_sem_seg':
+                PixelData() if only_prediction else PixelData(
```
Comment (MeowZheng) [inline mmseg/models/segmentors/base.py:195]: Is it necessary to add an empty PixelData for `gt_sem_seg` when only predict?

Comment (xiexinch) [reply mmseg/models/segmentors/base.py]: removed

**Review by MeowZheng — APPROVED** [review]
